### PR TITLE
SOLR-15992: Globally forbid and exclude known bad dependencies

### PR DESCRIPTION
**Pull request by risdenk**
https://issues.apache.org/jira/browse/SOLR-15992

### CODE REVIEW

**Comment by risdenk** [comment]
I'm planning to merge PR #612 first so only the one commit needs to be reviewed.

File: gradle/hacks/global-exclude-dependencies.gradle [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Globally exclude known bad dependencies to ensure they aren't included in any transitive dependency.
+
+allprojects { prj ->
+  plugins.withType(JavaPlugin) {
+    configurations.all {
```
Comment (dweiss) [inline gradle/hacks/global-exclude-dependencies.gradle:22]: Don't exclude it from all configurations. Exclude these from only those configurations that are created by the java plugin. I've talked about it a few times - you have no control over what configurations are defined by plugins; if you exclude arbitrary dependencies from these, you risk hours spending debugging why something in the build doesn't work as expected.

Comment (dsmiley) [reply gradle/hacks/global-exclude-dependencies.gradle:22]: I kind of wonder if it's worth the risk.  If the outcome is a ClassNotFoundException blahBlah.animalSnipper.Silliness then I think we'd find the issue quickly.

Comment (janhoy) [reply gradle/hacks/global-exclude-dependencies.gradle:22]: I'm not enough into how Gradle's build-time classpath (plugins etc) is isolated from the compile-time classpath that ends up in our binary distribution. The goal of these exclusions is to stop certain jars from ending up in our distro, so I'm +1 to anything that achieves that goal -- globally, without having to duplicate the exclusions everywhere.

Any updates needed in forbidden-apis?

Comment (dweiss) [reply gradle/hacks/global-exclude-dependencies.gradle:22]: Correct - don't use .all, use explicit names of configurations which are actually used to declare external dependencies (most of them will come from gradle's java plugin... you can define your own but typically this won't be the case).

Comment (dweiss) [reply gradle/hacks/global-exclude-dependencies.gradle:22]: > I kind of wonder if it's worth the risk. If the outcome is a ClassNotFoundException blahBlah.animalSniffer.Silliness then I think we'd find the issue quickly.

The outcome is unpredictable. Depends what the configuration you're excluding stuff from is used for. Try afterEvaluate {configurations.all { println it } } and you'll see how many there are in each project.

Comment (risdenk) [reply gradle/hacks/global-exclude-dependencies.gradle]: Ok I see what Dawid meant here. I fixed this up in ca16326ce3188915fef685b0cec57b4ce7562655

**Comment by dweiss** [comment]
I did take a look but it's not really helping me understand why it works or whether it's correct. :( Sorry. I can't even remember where we had exclusions on that custom configuration. If it does work, I'd commit it in.

**Comment by risdenk** [comment]
No worries Dawid - Appreciate the help and pointers!

@janhoy / @dsmiley any other comments/concerns/critiques/criticisms?

**Review by dsmiley — APPROVED** [review]

**Comment by risdenk** [comment]
Merged in 4a0854b64a82e29a822b262eab780ff849095392

**Pull request closed without merging**